records/README.md

```diff
@@ -0,0 +1,220 @@
+
+### The Interface
+
+The IPRS interface is below. It has a few types and functions. We use the Go language to express it, but this is language agnostic.
+
+```go
+// Record is the base type. user can define other types that
+// extend Record.
+type Record struct {
+  Scheme    Link // link to the validity scheme
+  Signature Link // link to a cryptographic signature over the rest of record
+  Value     Data // an opaque value
+}
+
+// Validator is a function that returns whether a record is valid.
+// Users provide their own Validator implementations.
+type Validator func(r *Record) (bool, error)
+
+// Marshal/Unmarshal specifies a way to code the record
+type Marshal(r *Record) ([]byte, error)
+type Unmarshal(r *Record, []byte) (error)
+```
+
+### Interface Example
+
+For example, Alice and Bob earlier could use the following interface:
+
+```protobuf
+type Record struct {
+  Scheme    Link // link to the validity scheme
+  Signature Link // link to a cryptographic signature over the record
+  Expires   Data // datetime at which record expires
+  Value     Data // an opaque value
+}
+
+
+func Validator(r *Record) (bool, error) {
+  authorKey := recordSigningKey(r)
+
+  // always check the signature first
+  sigok := authorKey.Verify(r.Signature, signablePart(r))
+  if !sigok {
+    return false, errors.New("invalid signature. forged record?")
+  }
+
+  // check the expiration.
+  if r.Expires < time.Now() {
+    return false, errors.New("record expired.")
+  }
+
+  return true, nil
+}
+
+func Marshal(r *Record) ([]byte, error) {
+  return recordProtobuf.Marshal(r)
+}
+
+func Unmarshal(r *Record, d []byte) (error) {
+  return recordProtobuf.Unmarshal(r, d)
+}
+```
+
+## Example Record Types
+
+For ease of use, IPRS implementations should include a set of common record types:
+
+- signed, valid within a datetime range
+- signed, expiring after a Time-To-Live
+- signed, based on ancestry (chain)
+- signed, with cryptographic freshness
+
+
+### Signed, valid within a datetime range
+
+This record type uses digital signatures (and thus a PKI) and timestamps (and thus a TI). It establishes that a record is valid during a particular datetime range. 0 (beginning of time), and infinity (end of time) can express unbounded validity.
+
+### Signed, expiring after a Time-To-Live
+
+This record type uses digital signatures (and thus a PKI) and TTLs (and thus a TI). It establishes that a record is valid for a certain amount of time after a particular event. For example, an event may be "upon receipt" to specify that a record is valid for a given amount of time after a processor first receives it. This is equivalent to the way DNS sets expiries.
+
+### Signed, based on ancestry (chain)
+
+This record type uses digital signatures (and thus a PKI) and merkle-links to other, previous records. It establishes that the "most recent" record (merkle-ordered) is the most valid. This functions similar to a git commit chain.
```

Comment:
On Fri, May 29, 2015 at 12:17:08PM -0700, Juan Batiz-Benet wrote:
> > Or is the request going to have special information (e.g. “give me
> > any records for that are descendents of ”, this seems to require
> > too much local storage on the record-serving node).
> 
> It may be possible to bundle with bitswap replication / query
> notation -- which will be something like git ref notation and
> xpath... -- (things like `want <path>/**/*` or `want <path>~10` OR
> go the full xpath route... oof.)

That sounds nice, but my main concern is that in the absence of
senders knowing what the requester will ask for they won't know which
auxiliary records to cache locally.  But I'm not familiar with bitswap
enough to say for sure.  Maybe you earn points by helping the
requester track down any records you're missing, and/or you benefit by
caching repeatedly-requested nodes follow-up objects locally.

This isn't a problem wit Git, etc., because the sending node always
has all the records available locally.  It's probably worth looking
into whether and how Git supports serving from a shallow clone, since
that's very similar to what we have in an IPRS node.